[FEATURE] Add a menu entry to memory layers to Make Permanent

nyalldawson opened the pull request:
Prompts for a location to save the layer to, then replaces it in place (keeping the same id, style, etc)

Goes best with a healthy serving of #7522  and #7516 

TODO: hide some of the vector file dialog settings which shouldn't be changed here, e.g. the field choices

![peek 2018-08-03 14-33](https://user-images.githubusercontent.com/1829991/43625336-d685aaf2-972f-11e8-940e-84920714fc3c.gif)

nyalldawson committed:
[FEATURE] Add a menu entry to memory layers to Make Permanent

Prompts for a location to save the layer to, then replaces it
in place (keeping the same id, style, etc)

nirvn commented:
This is fantastic. Please make sure that when you click on the memory indicator, that action will be trigger.

You'll end up curing me from my use of the memory layer saver plugin :-)

nyalldawson committed:
Spelling

nyalldawson committed:
Hide extra options from Make Permanent dialog

We don't want to e.g. allow users to change the CRS or drop
fields in this dialog.

nirvn commented on src/gui/ogr/qgsvectorlayersaveasdialog.h:
```diff
@@ -38,11 +38,18 @@ class GUI_EXPORT QgsVectorLayerSaveAsDialog : public QDialog, private Ui::QgsVec
+      SelectedOnly = 1 << 5, //!< Show selected features only option
+      GeometryType = 1 << 6, //!< Show geometry group
+      Extent = 1 << 7, //!< Show extent group
+      AllOptions = ~0 //!< Show all options
```
@nyalldawson , nice move.

nirvn commented on src/app/qgisapp.cpp:
```diff
@@ -7201,6 +7201,45 @@ void QgisApp::saveAsFile( QgsMapLayer *layer, bool onlySelected )
+      QgsDataProvider::ProviderOptions options;
+      vl->setDataSource( QStringLiteral( "%1" ).arg( newFilename ), vl->name(), QStringLiteral( "ogr" ), options );
+      this->messageBar()->pushMessage( tr( "Layer Saved" ),
+                                       tr( "Successfully saved scratch layer to <a href=\"%1\">%2</a>" ).arg( QUrl::fromLocalFile( QFileInfo( newFilename ).path() ).toString(), QDir::toNativeSeparators( newFilename ) ),
```
Don't use the path as hyperlink, use the file, it'll benefit from our recent native file opening work :)

nirvn commented on src/app/qgisapp.cpp:
```diff
@@ -7307,15 +7346,51 @@ void QgisApp::saveAsVectorFileGeneral( QgsVectorLayer *vlayer, bool symbologyOpt
+    }
+    this->emit layerSavedAs( vlayer, vectorFileName );
+    this->messageBar()->pushMessage( tr( "Layer Exported" ),
+                                     tr( "Successfully saved vector layer to <a href=\"%1\">%2</a>" ).arg( QUrl::fromLocalFile( QFileInfo( newFilename ).path() ).toString(), QDir::toNativeSeparators( newFilename ) ),
```
Also here, use file location instead of path location.

nyalldawson replied:
Thanks, both fixed

nirvn commented:
@nyalldawson , looking good, only two small comments that would be nice to address.

nyalldawson committed:
Fix message bar paths

nyalldawson committed:
Fix bad merge

nirvn commented:
@nyalldawson , thanks; +1 to merge. Nice side improvements on options to control visibility of settings in the dialog too.

nyalldawson committed:
Fix bad merge pt 2

nyalldawson committed:
Fix potential crash after exporting vector layer

nyalldawson committed:
Fix warning

nyalldawson committed:
Spelling
